### PR TITLE
Add file-source.

### PR DESCRIPTION
**Pull request by aartaka**
This adds a universally useful `file-source`. It searches for the file in the current directory, changing the directory properly.

# Things to Discuss
- Maybe add additional actions -- file deletion, touching, copying, opening in editor?
- Maybe add `C-l` for movement up the directories and other keybindings?

# How Has This Been Tested
I've tested it in my config with the function requiring file input. It worked.

How does this look? I'm still at the basic level of `prompter` understanding, but, hopefully, working on this will make me grok it :)

EDIT: It's fuzzy-sortable, yay!

### CODE REVIEW

**Comment by aartaka** [comment]
Fuzzy-matching works!

**Comment by aartaka** [comment]
- [ ] Another desirable thing: display only the basenames, without obvious clutter of absolute current directory name.

Example:
If the user are in the `/home/user/docs/` directory and are aware of that, they probably want to have a choice between `foo.txt`, `bar.md`, and `quux.org`, instead of `/home/user/docs/foo.txt`, `/home/user/docs/bar.md`, and `/home/user/docs/quux.org`.

File: libraries/prompter/prompter-source.lisp [hunk 1]
```diff
@@ -397,6 +397,30 @@ call.")))
...
+    (mapcar (lambda (file)
+              (make-instance 'suggestion
+                             :value file
+                             :match-data (namestring file)
+                             :properties (object-properties file)
+                             :source source
+                             :input input))
+            (append (uiop:subdirectories directory) (uiop:directory-files directory)))))
+
+;; TODO: `must-match-p'?
```
Comment (Ambrevar) [inline libraries/prompter/prompter-source.lisp:415]: I'm going to remove `must-match-p`, see #1213.

**Comment by Ambrevar** [comment]
This may be a starting point for #1219! ;)

I think we need to look at the old file manager code see what we can save.

Question: Why did you make the source a part of libraries/prompter?  Any specific reason?  True, it's quite general, but I'm concerned that a file manager is a lot of code (see Helm Find Files :p) and having this as part for a library may not ever be good enough.

So I'd suggest first moving it to a new file (e.g. `source/file-manager`), then adding `C-l`, actions, etc.

**Comment by Ambrevar** [comment]
About the basename display: Maybe you can find some inspiration in Helm Find Files. (Do you use Helm?)
So only display the basename and display the current working directory as part of the input.

**Comment by jmercouris** [comment]
This is really cool! Finally a file navigation that works! :-D

With this, it becomes quite feasible to actually use Nyxt as a Notepad replacement :-D

I agree with Pierre, my only comment is just about moving things to a new source file :-P

P.S. sorry about the repetition of my emoticons, I have yet to install the Kaomoji plugin :-O

**Comment by aartaka** [comment]
> About the basename display: Maybe you can find some inspiration in Helm Find Files. (Do you use Helm?)
> So only display the basename and display the current working directory as part of the input.

That's exactly what I want, and Helm is the source of inspiration there :D

**Comment by aartaka** [comment]
> This may be a starting point for #1219! ;)
>
> I think we need to look at the old file manager code see what we can save.

It actually is a starting point :)

> Question: Why did you make the source a part of libraries/prompter? Any specific reason? True, it's quite general, but I'm concerned that a file manager is a lot of code (see Helm Find Files :p) and having this as part for a library may not ever be good enough.

It's a suspicious practice to use a part of a particular mode everywhere. Modes should be as decoupled as possible. That's why putting this source in a mode is not a good idea -- it's already used in three other places/modes with more to come. 

We better ensure mode independence and put `file-source` either in a separate file in the core of Nyxt, or in the `prompter` itself (this I believe to be right, because it's as universally useful as `raw-input`). In any case, it can be extended and inherited afterwards, and `file-manager-mode` will most probably do that, much like `download-mode` does.

Having `open-file` in `download-mode` is confusing, so we better move it somewhere else.

> So I'd suggest first moving it to a new file (e.g. `source/file-manager`), then adding `C-l`, actions, etc.

Creating this file is all fine by me. Small nits, though:
- `file-source` should better be a part of `prompter` (see above).
- The `file-manager-mode` makes more sense as `prompt-buffer` mode, rather than a plain buffer mode. I can change my opinion on that when I get to implementation, though.
- Some actions (like file deletion) can already be part of `file-source`, as they only require UIOP, with no `file-manager-mode`-specific hacks.

**Comment by Ambrevar** [comment]
>> Question: Why did you make the source a part of libraries/prompter? Any specific reason? True, it's quite general, but I'm concerned that a file manager is a lot of code (see Helm Find Files :p) and having this as part for a library may not ever be good enough.
>
> It's a suspicious practice to use a part of a particular mode everywhere. Modes should be as decoupled as possible. That's why putting this source in a mode is not a good idea -- it's already used in three other places/modes with more to come. 

I didn't mean to make a mode, just to move it to `source`.

> Having `open-file` in `download-mode` is confusing, so we better move it somewhere else.

Agreed!

>> So I'd suggest first moving it to a new file (e.g. `source/file-manager`), then adding `C-l`, actions, etc.
>
> Creating this file is all fine by me. Small nits, though:
> - `file-source` should better be a part of `prompter` (see above).

I disagree: I think we should move it to `source` because it's a
non-trivial source that we will expand tremendously in the future.

**Comment by aartaka** [comment]
Sounds fine to me until it's not in a separate mode :)

**Comment by aartaka** [comment]
Okay, now there's `file-manager.lisp` in `source` (is there a better name?)

I have also extracted some helpers to simplify the code -- `directory-elements` and `ensure-directory-pathname`.

**Comment by aartaka** [comment]
Should be mergeable now.

**Comment by jmercouris** [comment]
Thank you Artyom. I too believe it is ready. If you like I can test it and merge.

Is there a better name? I was thinking `file-system.lisp`, though `file-manager.lisp` is also OK!

File: source/download-mode.lisp [hunk 2]
```diff
@@ -147,10 +147,11 @@ download."
   "Download the page or file of the current buffer."
   (download (current-buffer) (url (current-buffer))))
 
-(define-class downloaded-files-source (prompter:source)
-  ((prompter:name "Files")
-   (prompter:must-match-p t)
+(define-class downloaded-files-source (file-source)
+  ((prompter:must-match-p t)
    (prompter:constructor (mapcar #'destination-path (downloads *browser*)))
```
Comment (jmercouris) [inline source/download-mode.lisp:152]: I've thought that perhaps instead the files-source should have a slot with the path which determines the starting point.

File: source/file-manager.lisp [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+                             :value file
+                             :match-data (namestring file)
+                             :properties (prompter:object-properties file)
+                             :source source
+                             :input input))
+            (directory-elements (if (uiop:directory-pathname-p pathname)
+                                    pathname
+                                    (uiop:pathname-directory-pathname pathname))))))
+
+(define-class file-source (prompter:source)
```
Comment (jmercouris) [inline source/file-manager.lisp:26]: here is where the `path` slot would go representing the current path. Not entirely necessary, just a thought :-)

**Comment by jmercouris** [comment]
I have to say, I am REALLY digging the ivy like completion. This interface is actually better than my current emacs config! Very nice Artyom!!! Very nice!

**Comment by jmercouris** [comment]
Ah, one other thing, I would add a "Raw file" source so that the user can open new files. I think it should be enough to add a raw-source to the open-editor-with-new-file prompter instance.

here is what I am saying:
```
(define-command editor-open-file (&key (buffer (current-buffer)))
  "Open a file in the internal editor."
  (let ((file (first (prompt
                      :prompt "Open file"
                      :sources
                      (list (make-instance 'prompter:raw-source)
                            (make-instance 'file-source
                                           :name "Absolute file path"))))))
    (open-file buffer file)
    ;; TODO: Maybe make `editor-mode' and `editor-buffer' pathname-friendly?
    (setf (file buffer) (namestring file))
    (setf (title buffer) (namestring file))
    (setf (url buffer) (quri::make-uri-file :path (namestring file)))))
```

**Comment by jmercouris** [comment]
Please let me know if you are OK with these changes, we can always revert them!

**Comment by jmercouris** [comment]
If you are OK, then let's merge!

**Comment by aartaka** [comment]
Yes, that looks alright to me :)

**Comment by aartaka** [comment]
How about `files.lisp` instead of `file-manager.lisp`?

**Comment by aartaka** [comment]
`files.lisp` is too vague. Let it be `file-manager.lisp` for historical reasons :)

**Comment by jmercouris** [comment]
OK! Thank you once again Artyom!

**Pull request merged**